presto-hive/src/test/java/com/facebook/presto/hive/TestHiveRecoverableGroupedExecution.java

```diff
@@ -0,0 +1,250 @@
+                ListenableFuture<MaterializedResult> insert = executor.submit(() -> queryRunner.execute(
+                        recoverableSession,
+                        "INSERT INTO test_insert\n" +
+                                "SELECT key1, value1, 'foo'\n" +
+                                "FROM test_table1\n" +
+                                "JOIN test_table2\n" +
+                                "ON key1 = key2\n" +
+                                "JOIN test_table3\n" +
+                                "ON key2 = key3"));
+
+                // wait for queries to start running
+                while (getRunningQueryCount(queryRunner.getQueries()) != 2) {
+                    sleep(10);
+                }
+
+                // wait for additional 500ms before close a worker
+                sleep(500);
+                assertEquals(getRunningQueryCount(queryRunner.getQueries()), 2);
+                TestingPrestoServer worker = queryRunner.getServers().stream()
+                        .filter(server -> !server.isCoordinator())
+                        .findFirst()
+                        .get();
+                worker.close();
+
+                assertEquals(createTable.get(30, SECONDS).getUpdateCount(), OptionalLong.of(15000));
+                assertEquals(insert.get(30, SECONDS).getUpdateCount(), OptionalLong.of(15000));
+            }
+            finally {
+                queryRunner.execute(recoverableSession, "DROP TABLE IF EXISTS test_table1");
+                queryRunner.execute(recoverableSession, "DROP TABLE IF EXISTS test_table2");
+                queryRunner.execute(recoverableSession, "DROP TABLE IF EXISTS test_table3");
+                queryRunner.execute(recoverableSession, "DROP TABLE IF EXISTS test_create_table");
+                queryRunner.execute(recoverableSession, "DROP TABLE IF EXISTS test_insert");
+            }
+        }
+    }
+
+    private static long getRunningQueryCount(List<BasicQueryInfo> basicQueryInfos)
+    {
+        return basicQueryInfos.stream()
+                .map(BasicQueryInfo::getState)
+                .filter(queryState -> queryState == RUNNING)
+                .count();
+    }
+
+    private static Session createRecoverableSession(Optional<SelectedRole> role)
+    {
+        return testSessionBuilder()
+                .setIdentity(new Identity(
+                        "hive",
+                        Optional.empty(),
+                        role.map(selectedRole -> ImmutableMap.of("hive", selectedRole))
+                                .orElse(ImmutableMap.of())))
+                .setSystemProperty(COLOCATED_JOIN, "true")
+                .setSystemProperty(GROUPED_EXECUTION_FOR_AGGREGATION, "true")
+                .setSystemProperty(DYNAMIC_SCHEDULE_FOR_GROUPED_EXECUTION, "true")
+                .setSystemProperty(RECOVERABLE_GROUPED_EXECUTION, "true")
+                .setCatalogSessionProperty(HIVE_CATALOG, USE_REWINDABLE_SPLIT_SOURCE, "true")
```

Comment:
I guess we could remove this session property altogether. Let me do that in a separate commit.

Reply:
Since you removed the `USE_REWINDABLE_SPLIT_SOURCE`, you should enable rewindable execution for the tests in HiveIntegrationSmokeTests in a new way. Let's keep that path tested.

Reply:
Feel free to add it as a separate commit.